src/httpx/asyncstream.nim

```diff
@@ -0,0 +1,392 @@
+  
+  afterWriteCb: Option[AsyncStreamCb]
+    ## Optional callback to be called after every write.
+    ## Note that the write will already have been performed once this callback is called, and thus the queue will have already been modified.
+
+  isCompletedInternal: bool
+    ## Internal property, see isCompleted getter
+
+  exceptionInternal: Option[ref Exception]
+    ## Internal property, see exception getter
+
+proc newAsyncStream*[T](
+  maxQueueLen: range[1..high(int)],
+  afterReadCb: Option[AsyncStreamCb] = none[AsyncStreamCb](),
+  afterWriteCb: Option[AsyncStreamCb] = none[AsyncStreamCb](),
+): AsyncStream[T] =
+  ## Creates a new AsyncStream.
+  ## 
+  ## Arguments:
+  ##  - `maxQueueLen`: The maximum queue length to use
+  ##  - `afterReadCb`: Optional callback to be called after every read (defaults to None)
+  ##  - `afterWriteCb`: Optional callback to be called after every write (defaults to None)
+
+  result = new AsyncStream[T]
+  result.queue = initDeque[T](maxQueueLen)
+  result.maxQueueLenInternal = maxQueueLen
+  result.readCbs = initDeque[AsyncStreamCb](1)
+  result.writeCbs = initDeque[AsyncStreamCb](1)
+  result.afterReadCb = afterReadCb
+  result.afterWriteCb = afterWriteCb
+  result.isCompletedInternal = false
+  result.exceptionInternal = none[ref Exception]()
+
+func isCompleted*[T](this: AsyncStream[T]): bool {.inline.} =
+  ## Whether the stream is completed.
+  ## If true, exception will be None.
+  
+  return this.isCompletedInternal
+
+func isFailed*[T](this: AsyncStream[T]): bool {.inline.} =
+  ## Whether the stream is failed.
+  ## If true, exception will be Some.
+  
+  return this.exceptionInternal.isSome
+
+func isFinished*[T](this: AsyncStream[T]): bool {.inline.} =
+  ## Whether the stream is either completed or failed.
+  ## To check each individually, use isCompleted and isFailed, respectively.
+  
+  return this.isCompleted or this.isFailed
+
+func exception*[T](this: AsyncStream[T]): Option[ref Exception] {.inline.} =
+  ## The exception that was the cause of the stream failing to be written/read, or none if there was no exception.
+  ## If an exception is present, any reads to the stream when the queue is empty will raise this exception.
+  ## The isFinished property should be true if this is some.
+  
+  return this.exceptionInternal
+
+func queueLen*[T](this: AsyncStream[T]): int {.inline.} =
+  ## The current internal queue length
+
+  return this.queue.len
+
+func maxQueueLen*[T](this: AsyncStream[T]): int {.inline.} =
+  ## The maximum queue length
+  
+  return this.maxQueueLenInternal
+
+func queue*[T](this: AsyncStream[T]): Deque[T] {.inline.} =
+  ## The internal queue.
+  ## 
+  ## Warning: modifying the queue manually while there are pending reads or writes will cause a deadlock until a new read/write occurs.
+  ## Exercise caution when modifying it!
+
+proc `maxQueueLen=`*[T](this: AsyncStream[T], newLen: range[1..high(int)]) {.inline.} =
```

Comment:
Is `queue` supposed to have a function body here?

Reply:
Yes, it should definitely have a body. I'm guessing that might be the culprit of one of the really annoying bugs I was finding. Gimme a minute, lemme fix that

Reply:
@vabresto Thanks for catching that. Man, not enforcing a return value and having a default return value was such a mistake in terms of language design

Reply:
@ termermc Yeah, but there is an expermentatal feature (and fearture direction) to catch this error using definite assignment analysis => https://nim-lang.github.io/Nim/manual_experimental.html#strict-definitions-and-nimout-parameters.

Reply:
For example

```nim
{.experimental: "strictdefs".}
{.push warningAsError[ProveInit]: on.}

proc foo: int =
  ## docs
  ```
gives a proper error.

Reply:
@ringabout Thanks for letting me know. I'm going to be using that on new code from now on.